src/ImmersedBoundaries/immersed_grid_metrics.jl

```diff
@@ -34,6 +36,7 @@ end
 @inline Δzᵃᵃᶠ(i, j, k, ibg::IBG) = Δzᵃᵃᶠ(i, j, k, ibg.underlying_grid)
 
 return_metrics(grid::IBG) = return_metrics(grid.underlying_grid)
-xspacings(X, grid::IBG) = xspacings(X, grid.underlying_grid)
-yspacings(Y, grid::IBG) = yspacings(Y, grid.underlying_grid)
-zspacings(Z, grid::IBG) = zspacings(Z, grid.underlying_grid)
+
+xspacings(grid::IBG, args...; kwargs...) = xspacings(grid.underlying_grid, args...; kwargs...)
+yspacings(grid::IBG, args...; kwargs...) = yspacings(grid.underlying_grid, args...; kwargs...)
+zspacings(grid::IBG, args...; kwargs...) = zspacings(grid.underlying_grid, args...; kwargs...)
```

Comment:
We can proceed by dispatching this method only when valid, ie for `GridFittedBoundary` and `GridFittedBottom`. We'll need a different implementation for `PartialCellBottom`.

I think the right approach is to use `KernelFunctionOperation` for `xspacing`. Then every `spacings` method can simply collect the output of `xspacing`. This is generically valid and would presumably reduce the amount of code we have to write by a bit.

Reply:
oh good point!

so, for `PartialCellBottom` only the `zspacings` is not correct, right? the horizontal spacings don't change.

Reply:
Correct

Reply:
But it's possible to write code that works for all grids by referencing the kernel functions directly (rather than trying to extract the spacings from the grid properties). That's how we get spacings inside the code, so if we use the same method in the user interface then all is good.

Reply:
Agree that `KernelFunctionOperation`s would reduce the amount of code. It would also return an `AbstractOperation` instead of the current options (which are either a float or an array), which I think has extra computational costs, no? Also I think we'd lose a bit of the simplicity on the user side. For example the following line of code (taken from [this doc example](https://clima.github.io/OceananigansDocumentation/stable/generated/ocean_wind_mixing_and_convection/) would become way less readable:

```julia
lines!(ax, zspacings(grid, Center()), znodes(grid, Center()))
```

Finally, following the same logic, would `xnodes` etc. also become `KernelFunctionOperation`s?

Reply:
My suggestion is that `zspacing` returns an operation, and that `zspacings` returns an array that `collect`s the abstract operation. Something like

```julia
zspacings(grid, lx, ly, lz) = [zspacing(1, 1, k, grid, lx, ly, lz) for k = 1:Nz]
```

We don't want exactly the above. Maybe extending `collect` would work. We'd have to look into it.

For `zspacings` there is an extra cost, since for grids that don't modify their metrics (ie all grids except `ImmersedBoundaryGrid` with `PartialCellBottom`) we don't need to compute anything. But `zspacings` shouldn't be used within hot inner loops anyways --- its for post and pre processing, right? The cost is probably completely negliglble compared to the cost of actually running a simulation or, for example, displaying a plot. But we can definitely optimize performance for the cases that would accept it (ie if you aren't using partial cells) if things become a problem.

For `zspacing` I don't think there's a cost. We're leveraging dispatch, in simple cases we'd be calling down to functions like

https://github.com/CliMA/Oceananigans.jl/blob/87e3d770cf6b4502a891dbc587e1036968f36cb2/src/Operators/spacings_and_areas_and_volumes.jl#L104

So for the indexed `spacing` functions, the point is just to reuse that code rather than rewrite it.

> Finally, following the same logic, would xnodes etc. also become KernelFunctionOperations?

I think we could use the same simplifications for `xnodes` for sure. I'm not 100% sure if we should modify the node positions when we use partial cells. Probably we do... which would mean that it's not only simpler but necessary for `znodes`, etc.

Reply:
Hmm right I was a little confused. `zspacing` is actually the kernel function itself (not an operation or array).

For `zspacings` we do either have to (1) represent it with an operation or (2) compute the metrics. There's no way around this in general: with partial cells, we compute the metrics on the fly (this computation seems to have no cost on the GPU). For cut cells, computing the metrics on the fly is required in fact, because precomputing all the metrics (3D arrays) would greatly increase our memory footprint, and also probably incur a big cost since we are memory-bound.

We could implement optimized `zspacings` for non-immersed grids or `GridFittedBottom`. But is there a point? I don't think the costs would be noticeable compared to, for example, the computational cost of displaying a plot.

Reply:
> which would need to be first converted into a 3D Field, computed, and transformed into a 1D Array before plotting:

We have 1D fields. Also operations like "tranforming a field into an array" are totally negligible. For example,  we might do something like this O(100) times in a single time step.

I suggest we put performance to the side right now and think first about a clean, simple implementation. We can optimize performance later if needed, but this isn't performance critical code anyways.